widgets/url-sync/index.js

```diff
@@ -0,0 +1,172 @@
+    } else if (window.attachEvent) {
+      window.attachEvent('onhashchange', cb); 
+    }
+  },
+  pushState: function(qs){
+    window.location.assign(this.character + qs);
+  },
+  replaceState: function(qs){
+    window.location.replace(this.character + qs);
+  },
+  readUrl: function(){
+    return window.location.hash.slice(1);
+  }
+};
+
+/**
+ * Handles the modern API
+ * @type {UrlUtil}
+ */
+var modernUrlUtils = {
+  character: '?',
+  onpopstate: function(cb){
+    window.addEventListener('popstate', cb);
+  },
+  pushState: function(qs){
+    window.history.pushState(null, '', this.character + qs);
+  },
+  replaceState: function(qs){
+    window.history.replaceState(null, '', this.character + qs);
+  },
+  readUrl: function(){
+    return window.location.search.slice(1);
+  }
+};
+
+/**
+ * Instanciate a url sync widget. This widget let you synchronize the search
+ * parameters with the URL. It can operate with legacy API and hash or it can use
+ * the modern history API. By default, it will use the modern API, but if you are
+ * looking for compatibility with IE8 and IE9, then you should set 'useHash' to
+ * true.
+ * @class
+ * @param {object} urlUtils an object containing the function to read, watch the changes
+ * and update the URL.
+ * @param {number} threshold time in ms after which a new state is created in the browser
+ * history. The default value is 700.
+ * @param {string[]} trackedParameters parameters that will be synchronized in the
+ * URL. By default, it will track the query, all the refinable attribute (facets and numeric
+ * filters), the index and the page.
+ * @param {boolean} useHash if set to true, the url will be hash based. Otherwise,
+ * it'll use the query parameters using the modern history API.
+ */
+class URLSync {
+  constructor(urlUtils, options) {
+    this.__initLast = true;
+    this.urlUtils = urlUtils;
+    this.originalConfig = null;
+    this.timer = timerMaker(getNow());
+    this.threshold = options.threshold || 700;
+    this.trackedParameters = options.trackedParameters || ['query', 'attribute:*', 'index', 'page'];
+  }
+
+  getConfiguration(currentConfiguration) {
+    this.originalConfig = currentConfiguration;
+    var queryString = this.urlUtils.readUrl();
+    var config = AlgoliaSearchHelper.getConfigurationFromQueryString(queryString);
+    return config;
+  }
+  
+  init(state, helper) {
+    var self = this;
+    this.urlUtils.onpopstate(function() {
+      var qs = self.urlUtils.readUrl();
+      var partialState = AlgoliaSearchHelper.getConfigurationFromQueryString(qs);
+      var fullState = merge({}, self.originalConfig, partialState);
+      //compare with helper.state
+      var partialHelperState = helper.getState(self.trackedParameters);
+      var fullHelperState = merge({}, self.originalConfig, partialHelperState);
+
+      if(isEqual(fullHelperState, fullState)) return;
+
+      forEach(window.document.querySelectorAll('input'), function(i) {
+        i.blur();
```

Comment:
Yes this solution or specifically target the <inputs> we know, like in the searchBox, adding a data-ais-input-controlled/uncontrolled and then querySelector only on this.

We just need to avoid triggering blur() on things that are not ours because it's tricky.